Type mismatch between associated type and concrete type which are the same type.

Trying to build the following code:

```rust
trait WithLifetime<'a> {
    type Type;
}

struct Foo;

enum FooRef {}

impl<'a> WithLifetime<'a> for FooRef {
    type Type = &'a Foo;
}

fn wub<T, F>(f: F)
    where T: for<'a> WithLifetime<'a>,
          F: for<'a> FnOnce(&'a Foo) -> <T as WithLifetime<'a>>::Type
{
}

fn main() {
    wub::<FooRef, _>(|foo| foo)
}
```

Gives the error:

```
foo.rs:20:28: 20:31 error: mismatched types:
 expected `<FooRef as WithLifetime<'a>>::Type`,
    found `&'a Foo`
(expected associated type,
    found &-ptr) [E0308]
foo.rs:20     wub::<FooRef, _>(|foo| foo)
                                     ^~~
```

However `<FooRef as WithLifetime<'a>>::Type` and `&'a Foo` are the same type. And the compiler should be able to see that.

Triage: error message is different now:

```text
   Compiling playground v0.0.1 (/playground)
error[E0271]: type mismatch resolving `for<'a> <[closure@src/main.rs:20:22: 20:31] as std::ops::FnOnce<(&'a Foo,)>>::Output == <FooRef as WithLifetime<'a>>::Type`
  --> src/main.rs:20:5
   |
20 |     wub::<FooRef, _>(|foo| foo)
   |     ^^^^^^^^^^^^^^^^ expected &Foo, found associated type
   |
   = note: expected type `&Foo`
              found type `<FooRef as WithLifetime<'_>>::Type`
note: required by `wub`
  --> src/main.rs:13:1
   |
13 | / fn wub<T, F>(f: F)
14 | |     where T: for<'a> WithLifetime<'a>,
15 | |           F: for<'a> FnOnce(&'a Foo) -> <T as WithLifetime<'a>>::Type
16 | | {
17 | | }
   | |_^

error: aborting due to previous error
```

--- a/tests/ui/higher-ranked/hrtb-projection-closure-return-34430.rs
+++ b/tests/ui/higher-ranked/hrtb-projection-closure-return-34430.rs
@@ -1,0 +1,41 @@
+//! Regression test for <https://github.com/rust-lang/rust/issues/34430>.
+//!
+//! An associated-type projection under a higher-ranked binder
+//! (`for<'a> FnOnce(&'a Foo) -> <T as WithLifetime<'a>>::Type`) failed to normalize,
+//! so returning `&'a Foo` from the closure was rejected even though
+//! `<FooRef as WithLifetime<'a>>::Type` *is* `&'a Foo`.
+
+//@ revisions: current next
+//@ ignore-compare-mode-next-solver (explicit revisions)
+//@[next] compile-flags: -Znext-solver
+//@ check-pass
+
+#![allow(unused)]
+
+trait WithLifetime<'a> {
+    type Type;
+}
+
+struct Foo;
+
+enum FooRef {}
+
+impl<'a> WithLifetime<'a> for FooRef {
+    type Type = &'a Foo;
+}
+
+fn wub<T, F>(f: F)
+where
+    T: for<'a> WithLifetime<'a>,
+    F: for<'a> FnOnce(&'a Foo) -> <T as WithLifetime<'a>>::Type,
+{
+}
+
+fn main() {
+    wub::<FooRef, _>(|foo| foo);
+
+    // Annotating the closure's return type used to ICE instead. Both the concrete
+    // and the projected spelling are checked, since either could regress alone.
+    wub::<FooRef, _>(|foo| -> &Foo { foo });
+    wub::<FooRef, _>(|foo| -> <FooRef as WithLifetime>::Type { foo });
+}

